### PR TITLE
Support upcasters within async projections

### LINKED ISSUE #2615: Async Daemon does not take upcasters into consideration.
Mentioned on Discord. Upcasters are currently limited to inline aggregation and do not get applied for async projections or rebuilds.

### PR DESCRIPTION
**Pull request by Hawxy**
Closes #2615

### CODE REVIEW

File: src/EventSourcingTests/SchemaChange/Upcasters.cs [hunk 1]
```diff
@@ -734,14 +736,14 @@ private void Apply(ShoppingCartOpened @event)
     // Events in new namespace `New`
     namespace New
     {
-        public class ShoppingCart: AggregateBase
+        public class ShoppingCartAggregate: AggregateBase
```
Comment (oskardudycz) [inline src/EventSourcingTests/SchemaChange/Upcasters.cs:739]: **QUESTION:** @Hawxy, is there any reason for renaming this class? I'd prefer not to have suffixes in the type names, especially those that are used in the public documentation.

File: src/Marten/Events/Daemon/EventTypeFilter.cs [hunk 2]
```diff
@@ -9,18 +12,19 @@ namespace Marten.Events.Daemon;
...
         EventTypes = eventTypes;
-        _typeNames = eventTypes.Select(x => graph.EventMappingFor(x).Alias).ToArray();
+        // We need to load events that are mapped to this event type, not just the event itself.
+        var additionalAliases = graph.AliasesForEvents(eventTypes);
+        _typeNames = eventTypes.Select(x => graph.EventMappingFor(x).Alias).Union(additionalAliases).ToArray();
     }
 
-    public Type[] EventTypes { get; }
-
+    public IReadOnlyCollection<Type> EventTypes { get; }
```
Comment (oskardudycz) [inline src/Marten/Events/Daemon/EventTypeFilter.cs:27]: **QUESTION:** Could you expand on the change from Array to ReadonlyCollection?

**Review by oskardudycz — APPROVED** [review]
Looks good to me, just minor comments around the naming and API changes.

File: src/EventSourcingTests/SchemaChange/Upcasters.cs (hunk 1 above)
Comment (Hawxy) [reply src/EventSourcingTests/SchemaChange/Upcasters.cs]: Good point, has been fixed.

File: src/Marten/Events/Daemon/EventTypeFilter.cs (hunk 2 above)
Comment (Hawxy) [reply src/Marten/Events/Daemon/EventTypeFilter.cs:27]: Just avoids the `ToArray` when initialized from the projection base.

**Pull request merged**